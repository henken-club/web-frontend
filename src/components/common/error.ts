export type ErrorObject = {status: 400 | 404 | 500};
export type Serializable = {serialize(): ErrorObject};

export class TransformError extends Error implements Serializable {
  serialize(): ErrorObject {
    return {status: 400};
  }
}
export class BadRequestError extends Error implements Serializable {
  serialize(): ErrorObject {
    return {status: 404};
  }
}
export class ServerSideError extends Error implements Serializable {
  serialize(): ErrorObject {
    return {status: 500};
  }
}

export type ErrorHandling<T> = T | {error: ErrorObject};
