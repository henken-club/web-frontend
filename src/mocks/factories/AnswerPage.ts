import {
  AllAnswerPagesQuery,
  AllAnswerPagesQueryVariables,
  AnswerPageQuery,
  AnswerPageQueryVariables,
} from '../codegen';

import {
  alias,
  answerType,
  avatar,
  bookCover,
  comment,
  createdAt,
  displayName,
  id,
  repeat,
  title,
  which,
} from './common';

export const factoryAllAnswerPages = ({
  limit,
}: AllAnswerPagesQueryVariables): AllAnswerPagesQuery => ({
  __typename: 'Query',
  manyAnswers: repeat(limit, () => ({
    __typename: 'Answer',
    id: id(),
  })),
});

export const factoryAnswerPage = (
  variables: AnswerPageQueryVariables,
): AnswerPageQuery => ({
  __typename: 'Query',
  findAnswer: {
    __typename: 'FindAnswerPayload',
    answer: {
      __typename: 'Answer',
      id: id(),
      comment: comment(),
      createdAt: createdAt(),
      type: answerType(),
      answerTo: {
        __typename: 'Henken',
        id: id(),
        comment: comment(),
        postsTo: {
          __typename: 'User',
          id: id(),
          alias: alias(),
          displayName: displayName(),
          avatar: avatar(),
        },
        postedBy: {
          __typename: 'User',
          id: id(),
          alias: alias(),
          displayName: displayName(),
          avatar: avatar(),
        },
        content: which([
          {__typename: 'Book', id: id(), title: title(), cover: bookCover()},
          {__typename: 'BookSeries', id: id(), title: title()},
        ]),
      },
    },
  },
});
