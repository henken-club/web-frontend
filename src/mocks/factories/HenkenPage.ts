import {
  AllHenkenPagesQuery,
  AllHenkenPagesQueryVariables,
  HenkenPageQuery,
  HenkenPageQueryVariables,
} from '../codegen';

import {
  alias,
  authorName,
  avatar,
  bookCover,
  comment,
  displayName,
  id,
  repeat,
  title,
  which,
} from './common';

export const factoryAllHenkenPages = ({
  limit,
}: AllHenkenPagesQueryVariables): AllHenkenPagesQuery => ({
  __typename: 'Query',
  manyHenkens: {
    __typename: 'HenkenConnection',
    edges: repeat(limit, () => ({
      __typename: 'HenkenEdge',
      node: {
        __typename: 'Henken',
        id: id(),
      },
    })),
  },
});

export const factoryHenkenPage = (
  variables: HenkenPageQueryVariables,
): HenkenPageQuery => ({
  __typename: 'Query',
  findHenken: {
    __typename: 'FindHenkenPayload',
    henken: {
      __typename: 'Henken',
      id: variables.id,
      comment: comment(),
      postsTo: {
        __typename: 'User',
        id: id(),
        alias: alias(),
        displayName: displayName(),
        avatar: avatar(),
      },
      postedBy: {
        __typename: 'User',
        id: id(),
        alias: alias(),
        displayName: displayName(),
        avatar: avatar(),
      },
      content: which([
        {
          __typename: 'Book',
          id: id(),
          title: title(),
          cover: bookCover(),
        },
        {
          __typename: 'BookSeries',
          id: id(),
          title: title(),
        },
        {
          __typename: 'Author',
          id: id(),
          name: authorName(),
        },
      ]),
    },
  },
});
