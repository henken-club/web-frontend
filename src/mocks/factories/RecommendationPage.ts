import {
  AllRecommendationsPagesQuery,
  AllRecommendationsPagesQueryVariables,
  RecommendationPageQuery,
  RecommendationPageQueryVariables,
} from '../codegen';

import {
  alias,
  avatar,
  bookCover,
  displayName,
  score,
  id,
  title,
  updatedAt,
  which,
  repeat,
  authorName,
} from './common';

export const factoryAllRecommendationsPages = ({
  limit,
}: AllRecommendationsPagesQueryVariables): AllRecommendationsPagesQuery => ({
  __typename: 'Query',
  manyRecommendations: {
    __typename: 'RecommendationConnection',
    edges: repeat(limit, () => ({
      __typename: 'RecommendationEdge',
      node: {
        __typename: 'Recommendation',
        id: id(),
      },
    })),
  },
});

export const factoryRecommendationPage = (
  variables: RecommendationPageQueryVariables,
): RecommendationPageQuery => ({
  __typename: 'Query',
  findRecommendation: {
    __typename: 'FindRecommendationPayload',
    recommendation: {
      __typename: 'Recommendation',
      id: variables.id,
      score: score(),
      updatedAt: updatedAt(),
      recommendsTo: {
        __typename: 'User',
        id: id(),
        alias: alias(),
        displayName: displayName(),
        avatar: avatar(),
      },
      content: which([
        {__typename: 'Book', id: id(), title: title(), cover: bookCover()},
        {__typename: 'BookSeries', id: id(), title: title()},
        {__typename: 'Author', id: id(), name: authorName()},
      ]),
    },
  },
});
