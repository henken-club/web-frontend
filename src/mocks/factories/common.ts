import faker from 'faker';

import {AnswerType} from '../codegen';

// common
export const id = faker.datatype.uuid;
export const createdAt = () =>
  faker.date.between('2020-01-01', '2020-12-31').toISOString();
export const updatedAt = createdAt;

export const totalCount = faker.datatype.number;
export const hasNextPage = faker.datatype.boolean;

// Util
export const which = faker.random.arrayElement;
export const repeat = <T>(count: number, generator: () => T): T[] =>
  [...new Array(count)].map((_, i) => generator());
export const between = (min: number, max: number) =>
  min + faker.datatype.number(max - min);

// User
export const alias = faker.lorem.word;
export const displayName = faker.name.firstName;
export const avatar = () =>
  which([
    '/.mock/avatar_1.png',
    '/.mock/avatar_2.png',
    '/.mock/avatar_3.png',
    '/.mock/avatar_4.png',
    '/.mock/avatar_5.png',
    '/.mock/avatar_6.png',
    '/.mock/avatar_7.png',
    '/.mock/avatar_8.png',
    '/.mock/avatar_9.png',
    '/.mock/avatar_10.png',
    '/.mock/avatar_11.png',
    '/.mock/avatar_12.png',
  ]);

// Henken, Answer
export const title = faker.lorem.words;
export const comment = faker.lorem.words;

// Answer
export const answerType = () => which([AnswerType.Right, AnswerType.Wrong]);

// Recommendation
export const score = faker.datatype.number;

// Book
export const bookCover = () =>
  which([
    '/.mock/bookcover_1.jpg',
    '/.mock/bookcover_2.jpg',
    '/.mock/bookcover_3.jpg',
    '/.mock/bookcover_4.jpg',
    '/.mock/bookcover_5.jpg',
    '/.mock/bookcover_6.jpg',
    '/.mock/bookcover_7.jpg',
    '/.mock/bookcover_8.jpg',
  ]);

// Author
export const authorName = () => faker.name.findName();
