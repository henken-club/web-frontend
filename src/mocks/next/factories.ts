import faker from 'faker';

export const factoryUser = ({
  ...props
}: {
  id?: string;
  alias?: string;
  displayName?: string;
  avatar?: string;
} = {}) => ({
  __typename: 'User' as const,
  id: faker.datatype.uuid(),
  alias: faker.lorem.word(),
  displayName: faker.name.firstName(),
  avatar: faker.image.avatar(),
  ...props,
});

export const factoryUserEdge = ({
  node = [],
}: {node?: Parameters<typeof factoryUser>} = {}) => ({
  __typename: 'UserEdge' as const,
  node: factoryUser(...node),
});
