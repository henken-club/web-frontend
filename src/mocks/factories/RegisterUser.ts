import {RegisterUserMutation, RegisterUserMutationVariables} from '../codegen';

import {id} from './common';

export function factoryRegisterUser(
  variables: RegisterUserMutationVariables,
): RegisterUserMutation {
  return {
    __typename: 'Mutation',
    registerUser: {
      __typename: 'RegisterUserPayload',
      user: {__typename: 'User', id: id(), ...variables},
    },
  };
}
