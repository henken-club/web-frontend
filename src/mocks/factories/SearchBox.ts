import {SearchBoxQuery, SearchBoxQueryVariables} from '../codegen';

import {authorName, between, id, repeat, title, which} from './common';

export function factorySearchBox(
  variables: SearchBoxQueryVariables,
): SearchBoxQuery {
  return {
    __typename: 'Query',
    search: {
      __typename: 'SearchPayload',
      nodes: repeat(between(0, 5), () => ({
        __typename: 'SearchResult',
        content: which([
          {__typename: 'Author', id: id(), name: authorName()},
          {
            __typename: 'Book',
            id: id(),
            title: title(),
            writings: {
              __typename: 'WritingConnection',
              edges: repeat(between(1, 4), () => ({
                __typename: 'WritingEdge',
                node: {
                  __typename: 'Writing',
                  author: {
                    __typename: 'Author',
                    id: id(),
                    name: authorName(),
                  },
                },
              })),
            },
          },
          {__typename: 'BookSeries', id: id(), title: title()},
        ]),
      })),
    },
  };
}
