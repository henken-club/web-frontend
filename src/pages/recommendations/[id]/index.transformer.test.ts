import faker from 'faker';

import {RecommendationPageQuery} from './index.page.codegen';
import {transformContent, transformer} from './index.transform';

import {id} from '~/mocks/factories/common';
import {factoryRecommendationPage} from '~/mocks/factories/RecommendationPage';

describe('Recommendation Page transformer', () => {
  describe('transformContent', () => {
    it('Book if cover exists', () => {
      const actual = transformContent({
        __typename: 'Book',
        id: 'id',
        title: 'title',
        cover: 'cover',
      });

      expect(actual).toStrictEqual({
        type: 'Book',
        book: {
          id: 'id',
          title: 'title',
          cover: 'cover',
        },
      });
    });

    it('Book if cover exists', () => {
      const actual = transformContent({
        __typename: 'Book',
        id: 'id',
        title: 'title',
        cover: null,
      });

      expect(actual).toStrictEqual({
        type: 'Book',
        book: {
          id: 'id',
          title: 'title',
          cover: null,
        },
      });
    });

    it('BookSeries', () => {
      const actual = transformContent({
        __typename: 'BookSeries',
        id: 'id',
        title: 'title',
      });

      expect(actual).toStrictEqual({
        type: 'BookSeries',
        bookSeries: {
          id: 'id',
          title: 'title',
        },
      });
    });
  });

  describe('transform', () => {
    it('recommendationが存在しない場合', () => {
      const actual = transformer({
        __typename: 'Query',
        findRecommendation: {
          __typename: 'FindRecommendationPayload',
          recommendation: null,
        },
      });
      expect(actual).toBeNull();
    });

    it.each(
      [...new Array(10)].map((_, i): [RecommendationPageQuery] => {
        faker.seed(i);
        return [factoryRecommendationPage({id: id()})];
      }),
    )('recommendationが存在する場合 %#', (payload) => {
      const actual = transformer(payload);
      expect(actual).toMatchSnapshot();
    });
  });
});
