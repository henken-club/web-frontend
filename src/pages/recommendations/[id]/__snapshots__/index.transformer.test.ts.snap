// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Recommendation Page transformer transform recommendationが存在する場合 0 1`] = `
Object {
  "recommendation": Object {
    "content": Object {
      "bookSeries": Object {
        "id": "7aff1a3a-2fa9-4467-b392-51aa52c3f5ad",
        "title": "perferendis aspernatur fuga",
      },
      "type": "BookSeries",
    },
    "id": "89bd9d8d-69a6-474e-8f46-7cc8796ed151",
    "recommendsTo": Object {
      "alias": "blanditiis",
      "avatar": "https://cdn.fakercloud.com/avatars/logorado_128.jpg",
      "displayName": "Curt",
      "id": "5dfc2ddf-7cc7-48ca-9ba9-28fc816742cb",
    },
    "score": 64817,
    "updatedAt": "2020-01-08T09:06:47.434Z",
  },
}
`;

exports[`Recommendation Page transformer transform recommendationが存在する場合 1 1`] = `
Object {
  "recommendation": Object {
    "content": Object {
      "bookSeries": Object {
        "id": "ae3e4479-0999-42f9-8b31-86fb7680c80a",
        "title": "eos sint autem",
      },
      "type": "BookSeries",
    },
    "id": "6fbe024f-2316-4265-a6e8-d65a837e308a",
    "recommendsTo": Object {
      "alias": "a",
      "avatar": "https://cdn.fakercloud.com/avatars/koridhandy_128.jpg",
      "displayName": "Hannah",
      "id": "7862f3cc-bfc5-41b8-aded-1d0420ea196a",
    },
    "score": 91396,
    "updatedAt": "2020-06-01T07:35:24.309Z",
  },
}
`;

exports[`Recommendation Page transformer transform recommendationが存在する場合 2 1`] = `
Object {
  "recommendation": Object {
    "content": Object {
      "bookSeries": Object {
        "id": "1f5dd863-0b36-41ef-bf8c-b9bc72f46855",
        "title": "omnis quae recusandae",
      },
      "type": "BookSeries",
    },
    "id": "620e8f67-6552-43b9-9474-a9d8a27892bc",
    "recommendsTo": Object {
      "alias": "rerum",
      "avatar": "https://cdn.fakercloud.com/avatars/joreira_128.jpg",
      "displayName": "Madeline",
      "id": "876d21f8-4146-4112-a799-331c39527537",
    },
    "score": 25849,
    "updatedAt": "2020-11-07T16:49:24.774Z",
  },
}
`;

exports[`Recommendation Page transformer transform recommendationが存在する場合 3 1`] = `
Object {
  "recommendation": Object {
    "content": Object {
      "bookSeries": Object {
        "id": "318d1b77-55b8-4b7b-8b55-da7589d58720",
        "title": "corporis et aut",
      },
      "type": "BookSeries",
    },
    "id": "81bd4189-e6e0-4203-b017-b0277a345a79",
    "recommendsTo": Object {
      "alias": "ut",
      "avatar": "https://cdn.fakercloud.com/avatars/shinze_128.jpg",
      "displayName": "Rahul",
      "id": "c8f46684-4b97-4427-8ccd-443e6be0f2a5",
    },
    "score": 91330,
    "updatedAt": "2020-01-09T18:04:52.668Z",
  },
}
`;

exports[`Recommendation Page transformer transform recommendationが存在する場合 4 1`] = `
Object {
  "recommendation": Object {
    "content": Object {
      "bookSeries": Object {
        "id": "53ac3271-2aef-465d-a6be-56e569ddbdb9",
        "title": "libero dicta beatae",
      },
      "type": "BookSeries",
    },
    "id": "fe82fdb9-b932-4f30-b4e6-ac039d5f9269",
    "recommendsTo": Object {
      "alias": "occaecati",
      "avatar": "https://cdn.fakercloud.com/avatars/antonyryndya_128.jpg",
      "displayName": "Devin",
      "id": "9640afc6-0f0c-4fd7-a711-922dbd6e8dfd",
    },
    "score": 81194,
    "updatedAt": "2020-01-04T06:43:05.476Z",
  },
}
`;

exports[`Recommendation Page transformer transform recommendationが存在する場合 5 1`] = `
Object {
  "recommendation": Object {
    "content": Object {
      "bookSeries": Object {
        "id": "710e1417-40f4-4a3d-b96a-6c4e3ebde1a5",
        "title": "nihil aperiam similique",
      },
      "type": "BookSeries",
    },
    "id": "30dd35ef-7196-4c58-b4f3-c1abd732ce14",
    "recommendsTo": Object {
      "alias": "cumque",
      "avatar": "https://cdn.fakercloud.com/avatars/renbyrd_128.jpg",
      "displayName": "Rosalee",
      "id": "14ea8979-b404-414d-9d2e-2ff8f13f073a",
    },
    "score": 20415,
    "updatedAt": "2020-05-31T04:41:55.747Z",
  },
}
`;

exports[`Recommendation Page transformer transform recommendationが存在する場合 6 1`] = `
Object {
  "recommendation": Object {
    "content": Object {
      "bookSeries": Object {
        "id": "48327c59-b336-458d-898e-7fade2796746",
        "title": "rerum quod voluptatibus",
      },
      "type": "BookSeries",
    },
    "id": "ef53d105-1f91-4896-8579-d7cb68392abf",
    "recommendsTo": Object {
      "alias": "aliquam",
      "avatar": "https://cdn.fakercloud.com/avatars/myastro_128.jpg",
      "displayName": "Paris",
      "id": "76aecd70-ab2c-45b9-b789-11f3633db7f8",
    },
    "score": 91419,
    "updatedAt": "2020-10-26T05:57:28.097Z",
  },
}
`;

exports[`Recommendation Page transformer transform recommendationが存在する場合 7 1`] = `
Object {
  "recommendation": Object {
    "content": Object {
      "bookSeries": Object {
        "id": "69aa8466-0a17-4bc8-abbf-7af024a983df",
        "title": "qui rerum ducimus",
      },
      "type": "BookSeries",
    },
    "id": "13c57fb7-f484-4811-a407-8adc5661a4ae",
    "recommendsTo": Object {
      "alias": "quis",
      "avatar": "https://cdn.fakercloud.com/avatars/johnsmithagency_128.jpg",
      "displayName": "Vallie",
      "id": "57ee40c9-4fd3-428c-af27-86c5a3713779",
    },
    "score": 69403,
    "updatedAt": "2020-03-18T21:15:20.517Z",
  },
}
`;

exports[`Recommendation Page transformer transform recommendationが存在する場合 8 1`] = `
Object {
  "recommendation": Object {
    "content": Object {
      "book": Object {
        "cover": null,
        "id": "af452614-ac1a-471a-92ca-007ec3f99402",
        "title": "ducimus autem amet",
      },
      "type": "Book",
    },
    "id": "d0f3d68d-3609-46c6-a8d7-e828bc6bd906",
    "recommendsTo": Object {
      "alias": "vel",
      "avatar": "https://cdn.fakercloud.com/avatars/quailandquasar_128.jpg",
      "displayName": "Keyon",
      "id": "7fd50341-4fa2-4452-9738-6fe85af10554",
    },
    "score": 35281,
    "updatedAt": "2020-04-15T12:17:50.110Z",
  },
}
`;

exports[`Recommendation Page transformer transform recommendationが存在する場合 9 1`] = `
Object {
  "recommendation": Object {
    "content": Object {
      "book": Object {
        "cover": null,
        "id": "a39dce0c-9517-4296-b972-6edf2eb6b4e5",
        "title": "eaque maiores est",
      },
      "type": "Book",
    },
    "id": "05877025-203e-4623-a165-725e5fb0fbb9",
    "recommendsTo": Object {
      "alias": "ut",
      "avatar": "https://cdn.fakercloud.com/avatars/calebogden_128.jpg",
      "displayName": "Will",
      "id": "2af88bf6-ab9d-4a7c-b6cd-f12181f7a3de",
    },
    "score": 13542,
    "updatedAt": "2020-11-23T18:32:32.654Z",
  },
}
`;
